### PR TITLE
odpic: init at 2.3.2

### PR DESCRIPTION
**Pull request by mulderr**
###### Motivation for this change

Adds Oracle Database Programming Interface for C ([ODPI-C](https://oracle.github.io/odpi/)) library. 

Should also allow to fix build of [odpic-raw](http://hackage.haskell.org/package/odpic-raw) (Haskell bindings) which is currently failing due to missing dependency:

```
Setup: Missing dependency on a foreign library:
  * Missing C library: odpic
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by mulderr** [comment]
@pesterhazy @flokli As maintainers of oracle-instantclient you may also be interested in this.

**Comment by flokli** [comment]
@mulderr funny thing, odpi also comes as a submodule of cx_python (IIRC). Maybe you want to apply the patch from there here as well, or even use this separate odpi drv for everything.

**Comment by mulderr** [comment]
[This one](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/python-modules/cx_oracle) right?

Documentation says that ODPI-C can be used in two different ways. As a shared library or by including the source directly in another project. I think `cx_Oracle` does the latter while `odpic-raw` currently expects the former.

I think the patch is just fixing the path to `libclntsh`? I decided to override RPATH instead. The net effect should be identical but browsing through nixpkgs patchelf seems to be the goto solution.

Hm, while I don't have immediate use for it (other than to make `odpic-raw` work), it makes sense to me that this should be offered as a separate library in case anyone wants to use it from C/C++ etc.

Luckily, correct me if I'm wrong, this shouldn't break cx_Oracle in any way? So any project that was using odpic directly can keep using a version independent of the one in nixpkgs - for example via git submodules as cx_Oracle does.

**Comment by flokli** [comment]
Overriding rpath is also possible, and probably cleaner than maintaining a patch in nixpkgs. Agreed :-)

Since https://github.com/oracle/python-cx_Oracle/commit/9baa55b3bffaa1d79c443dd0f7db16162a22cb93, it's possible to build `cx_oracle` with an external odpic library, so I updated my drv accordingly to use the new odpi drv:

https://github.com/flokli/nixpkgs/commit/cx_oracle-odpi-c

Could you `git am` that on top of your PR?

File: pkgs/development/libraries/odpic/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    sha256 = "0a52vqy3c27hmcia4x4vhs6z0ha27xkw7fvfa17f7kxs5w84ll07";
+  };
+
+  buildInputs = [ libaio oracle-instantclient ];
+
+  libPath = stdenv.lib.makeLibraryPath
+    [ oracle-instantclient ];
+
+  dontPatchELF = true;
+  installPhase = ''
```
[the remark this answers is not in the document]
Comment (mulderr) [reply pkgs/development/libraries/odpic/default.nix]: I'd like to point out that 2.3.2 didn't even have an install target ;) Good call though. I'll update this to 2.4.0.

File: pkgs/development/libraries/odpic/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  buildInputs = [ libaio oracle-instantclient ];
+
+  libPath = stdenv.lib.makeLibraryPath
+    [ oracle-instantclient ];
+
+  dontPatchELF = true;
+  makeFlags = [ "PREFIX=$(out)" ];
+
+  postFixup = ''
+    patchelf --set-rpath "${libPath}" $out/lib/libodpic.so
```
Comment (FRidh) [inline pkgs/development/libraries/odpic/default.nix]: `stdenv.hostPlatform.extensions.sharedLibrary` instead of `so`

**Comment by flokli** [comment]
speaking of `stdenv.hostPlatform.extensions.sharedLibrary` and portability:

Took the opportunity to wire darwin support into the oracle-instantclient and odpic drvs in 7e749e1c557d2226a412ef10c05056e555a935b7 and fb5321f46aa2c553d669d99d6f1f6e2614147738.

@matthewbauer given that's my first darwin contribution, could you have a [closer look](https://github.com/flokli/nixpkgs/commits/oracle-instantclient-darwin)? 

Whats your opinion on the makefile patch for odpi, and what about `fixDarwinDylibNames` not picking up `$out/lib/libodpic.dylib.2.4.0` automatically (as `libodpic.dylib` is a symlink and thus skipped)?

Anyways, with this PR plus above mentioned two commits, I was able to use sqlalchemy to connect to oracle databases from a Mac with Nix installed, too :-)

**Comment by flokli** [comment]
@mulderr patches for odpic were upstreamed (https://github.com/oracle/odpi/issues/65),
I updated my [oracle-instantclient-darwin](https://github.com/flokli/nixpkgs/commits/oracle-instantclient-darwin) branch to include the patch.

Would you mind cherry-picking my commits in here?

**Comment by mulderr** [comment]
@flokli Done.

Fair warning though, I don't use Macs or have one handy to test so it's impossible for me to verify. If it works for you then cool. Might as well take care of this right away.

**Comment by flokli** [comment]
Umm, forgot to drop `0001-Makefile-fix-dylib-names.patch` from my commit, sorry :-/

If you want to test, you can try the AndrewDryga vagrant-box-osx.

**Comment by mulderr** [comment]
No worries. Patch removed.

**Comment by flokli** [comment]
LGTM. @matthewbauer, @FRidh, merge?

**Comment by mulderr** [comment]
Thanks for all the input and for merging! Awesome.

**Pull request merged**